src/publishing/s3publisher.py

```diff
@@ -168,24 +169,30 @@ def publish_to_s3(directory, base_url, site_prefix, bucket, federalist_config,
 
     # Upload new and replacement files
     upload_objects = new_objects + replacement_objects
-    for file in upload_objects:
-        if dry_run:  # pragma: no cover
+
+    # task to be run in parallel via threadpool
+    def uploader_task(client, file):
+        logger.info(f'Uploading {file.s3_key}')
+        try:
+            file.upload_to_s3(bucket, client)
+        except UnicodeEncodeError as err:
+            if err.reason == 'surrogates not allowed':
+                logger.warning(
+                    f'... unable to upload {file.filename} due '
+                    f'to invalid characters in file name.'
+                )
+            else:
+                raise
+
+    if dry_run:  # pragma: no cover
+        for file in upload_objects:
             logger.info(f'Dry-run uploading {file.s3_key}')
-        else:
-            logger.info(f'Uploading {file.s3_key}')
-
-            try:
-                file.upload_to_s3(bucket, s3_client)
-            except UnicodeEncodeError as err:
-                if err.reason == 'surrogates not allowed':
-                    logger.warning(
-                        f'... unable to upload {file.filename} due '
-                        f'to invalid characters in file name.'
-                    )
-                else:
-                    raise
-
-    # Delete files not needed any more
+    else:
+        with ThreadPoolExecutor(max_workers=8) as executor:
```

Comment:
Did you play with the `max_workers` to test performance? Should we turn it into a parameter to facilitate performance testing?

Reply:
I only tried this single value. In other contexts, I've seen that 25-50 gets max performance but happy to at least make this a constant for easier testing